### PR TITLE
Add support of configurable Ion encoding format.

### PR DESCRIPTION
**Pull request by linlin-s**
### Description
Added support for configurable Ion encoding formats, allowing users to specify either `Text` or `Binary` encoding when creating Ion format tables with Hive connector. By default, `Binary` encoding will be used if no format is specified.

### Changes
- Added new table property `ion.encoding` to control Ion encoding format. Supported values `Text` and `Binary`, the default value is Binary.
- Added tests to verify that the data written with binary or text encoding is readable by the pageSource.

### Usage
Users can now specify the Ion encoding format when creating tables with the Hive connector:
```
CREATE TABLE example (
  request_time TIMESTAMP,
  url VARCHAR,
) 
WITH (
    format = 'ION',
    ion.encoding = 'Text'
)

### CODE REVIEW

**Comment by linlin-s** [comment]
The workflow task run failure is from checking the first commit [868a2e25597a2fabec30ad20e3d54f5e8b104c4e](https://github.com/rmarrowstone/trino/pull/4/commits/868a2e25597a2fabec30ad20e3d54f5e8b104c4e). The error has been fixed in the second one.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/SerdeConstants.java [hunk 1]
```diff
@@ -28,5 +28,9 @@ public final class SerdeConstants
     public static final String LIST_COLUMN_TYPES = "columns.types";
     public static final String LIST_COLUMN_COMMENTS = "columns.comments";
 
+    public static final String ION_ENCODING = "ion.encoding";
+    public static final String TEXT_ENCODING = "Text";
+    public static final String BINARY_ENCODING = "Binary";
```
Comment (desaikd) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/SerdeConstants.java]: We should probably keep all the Ion related constants in `IonFileWriterFactory.java` or `IonFileWriter.java` where it is used. This file seems to contain more common constants for hive serde.

**Comment by rmarrowstone** [comment]
IIUC the hive-serde took all uppercase encoding names: https://docs.aws.amazon.com/athena/latest/ug/ion-serde-using-ctas-and-insert-into-to-create-ion-tables.html 

I prefer what you have here, so I think the best UX is just to make it case insensitive.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/ion/IonFileWriterFactory.java [hunk 2]
```diff
@@ -95,12 +100,17 @@ public Optional<FileWriter> createFileWriter(
                     .mapToObj(ordinal -> new Column(fileColumnNames.get(ordinal), fileColumnTypes.get(ordinal), ordinal))
                     .toList();
 
+            IonWriterBuilder writerBuilder = getIonEncoding(schema).equals(TEXT_ENCODING)
```
Comment (rmarrowstone) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/ion/IonFileWriterFactory.java]: Prefer to fail on a malformed or unsupported encoding option. e.g. in future we may want to support `JSON` but we don't today so that should hard fail; it would be surprising if instead it wrote binary ion.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestHiveFileFormats.java [hunk 3]
```diff
@@ -384,6 +386,24 @@ public void testIon(int rowCount, long fileSizePadding)
                 .isReadableByPageSource(fileSystemFactory -> new IonPageSourceFactory(fileSystemFactory));
     }
 
+    @Test(dataProvider = "validRowAndFileSizePadding")
+    public void testIonWithTextEncoding(int rowCount, long fileSizePadding)
```
Comment (rmarrowstone) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestHiveFileFormats.java:397]: Having this test doesn't hurt, but I think we need a test (in `IonPageSourceFactorySmokeTest` or a new derivative) that proves text writes, well text, and binary writes binary.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/ion/IonFileWriterFactory.java [hunk 4]
```diff
@@ -95,12 +100,17 @@ public Optional<FileWriter> createFileWriter(
...
+                    ? IonTextWriterBuilder.minimal()
+                    : IonBinaryWriterBuilder.standard();
+
```
Comment (desaikd) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/ion/IonFileWriterFactory.java]: Can we just pass the encoding flag to `IonFileWriter` and let it create the builder object instead of passing it from here?

**Review by rmarrowstone — COMMENTED** [review]
Looks like you have it working for the two happy paths. Let's cover the failure case and ensure we have solid tests.

**Comment by rmarrowstone** [comment]
> The workflow task run failure is from checking the first commit [868a2e25597a2fabec30ad20e3d54f5e8b104c4e](https://github.com/rmarrowstone/trino/pull/4/commits/868a2e25597a2fabec30ad20e3d54f5e8b104c4e). The error has been fixed in the second one.

The trino folks want every commit within a PR to be rollback-toable. So I think the preferred fix is to squash and re-push with force.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/SerdeConstants.java (hunk 1 above)
Comment (rmarrowstone) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/SerdeConstants.java]: We're also gonna end up with a whole bunch of these.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/ion/IonFileWriterFactory.java (hunk 2 above)
Comment (linlin-s) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/ion/IonFileWriterFactory.java]: I agree, should have the case to handle the unsupported option.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/ion/IonFileWriterFactory.java (hunk 4 above)
Comment (linlin-s) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/ion/IonFileWriterFactory.java]: Yeah, we can pass the flag and create the writers in `IonFileWrite`. It sounds better since it will reduce the unnecessary builder creation process.

File: lib/trino-hive-formats/src/main/java/io/trino/hive/formats/ion/IonConstants.java
Comment (linlin-s) [inline lib/trino-hive-formats/src/main/java/io/trino/hive/formats/ion/IonConstants.java:1]: The IonConstants includes all the constants related to Ion and they are used in plugin/hive. Should we move this class to` trino/plugin/hive`. The only reason to put under `lib/trino-hive-formats` is to align with other formats. i.e. [json](https://github.com/rmarrowstone/trino/tree/master/lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/json) or [csv](https://github.com/rmarrowstone/trino/tree/master/lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/csv)

**Comment by linlin-s** [comment]
> > The workflow task run failure is from checking the first commit [868a2e25597a2fabec30ad20e3d54f5e8b104c4e](https://github.com/rmarrowstone/trino/pull/4/commits/868a2e25597a2fabec30ad20e3d54f5e8b104c4e). The error has been fixed in the second one.
> 
> The trino folks want every commit within a PR to be rollback-toable. So I think the preferred fix is to squash and re-push with force.

I will `squash and merge` when I merge this PR.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/ion/IonFileWriterFactory.java [hunk 5]
```diff
@@ -100,17 +99,19 @@ public Optional<FileWriter> createFileWriter(
                     .mapToObj(ordinal -> new Column(fileColumnNames.get(ordinal), fileColumnTypes.get(ordinal), ordinal))
                     .toList();
 
-            IonWriterBuilder writerBuilder = getIonEncoding(schema).equals(TEXT_ENCODING)
-                    ? IonTextWriterBuilder.minimal()
-                    : IonBinaryWriterBuilder.standard();
+            Boolean binaryEncoding = switch (getIonEncoding(schema).toLowerCase(Locale.ROOT)) {
+                case TEXT_ENCODING -> true;
+                case BINARY_ENCODING -> false;
+                default -> throw new IllegalArgumentException("Unsupported Ion encoding format");
```
Comment (desaikd) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/ion/IonFileWriterFactory.java]: Maybe we should use the `TrinoException` here.
```suggestion
                default -> throw new TrinoException(HIVE_UNSUPPORTED_FORMAT, "Unsupported Ion encoding format");
```

File: lib/trino-hive-formats/src/main/java/io/trino/hive/formats/ion/IonConstants.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+package io.trino.hive.formats.ion;
+
+import java.util.Map;
+
+public final class IonConstants
+{
+    public static final String ION_ENCODING = "ion.encoding";
+    public static final String TEXT_ENCODING = "text";
+    public static final String BINARY_ENCODING = "binary";
+    public static final String ERROR_ENCODING = "error_encoding";
```
Comment (desaikd) [inline lib/trino-hive-formats/src/main/java/io/trino/hive/formats/ion/IonConstants.java]: Is `ERROR_ENCODING` only used for tests? If yes, then we should move it into some test util file.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/ion/IonFileWriterFactory.java (hunk 5 above)
Comment (linlin-s) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/ion/IonFileWriterFactory.java]: Will update this in the next commit. Thanks

File: lib/trino-hive-formats/src/main/java/io/trino/hive/formats/ion/IonConstants.java (hunk 6 above)
Comment (linlin-s) [reply lib/trino-hive-formats/src/main/java/io/trino/hive/formats/ion/IonConstants.java]: Yes, it is only used for the test. I'll move it to the test util file.

Comment (rmarrowstone) [reply lib/trino-hive-formats/src/main/java/io/trino/hive/formats/ion/IonConstants.java]: Yea definitely don't include something only for testing here.

File: plugin/trino-hive/pom.xml [hunk 7]
```diff
@@ -545,6 +545,13 @@
             <scope>test</scope>
         </dependency>
 
+        <dependency>
```
Comment (rmarrowstone) [inline plugin/trino-hive/pom.xml]: I'm thinking this will not be a welcome addition. The test _should_ all be using junit 5, aka "jupiter" now, no?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/ion/IonFileWriter.java [hunk 8]
```diff
@@ -51,6 +52,7 @@ public IonFileWriter(
             Closeable rollbackAction,
             TypeManager typeManager,
             Optional<CompressionKind> compressionKind,
+            Boolean encoding,
```
Comment (rmarrowstone) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/ion/IonFileWriter.java]: nit: variable name is ambiguous as a boolean.

I don't know about the trino code style in regards to this, but in general it's better to have a clear type for such flags as a method parameter. e.g. an `IonEncoding` enum, you could even put the logic of what writer to create as a method of the enum. That may be overengineering it, but at the least the name needs to change.

File: plugin/trino-hive/pom.xml (hunk 7 above)
Comment (linlin-s) [reply plugin/trino-hive/pom.xml]: Thanks for catching this. This was recommended by the dependency-analyze task in the build process. I rerun the build without this dependency, and the build was succeed. This should be removed.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/ion/IonWriterOptions.java
Comment (linlin-s) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/ion/IonWriterOptions.java:1]: Replace `IonConstants` with `IonWriterOptions`. We will include Ion serialization related properties in this class in the future. i.e.` ion.timestamp.serialization_offset`

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/ion/IonPageSourceSmokeTest.java [hunk 9]
```diff
@@ -346,11 +346,7 @@ private static PageSourceParameters preparePageSourceParameters(TrinoFileSystemF
                 length,
                 nowMillis);
 
-        final Map<String, String> tableProperties = ImmutableMap.<String, String>builder()
-                .put(LIST_COLUMNS, tableColumns.stream().map(HiveColumnHandle::getName).collect(Collectors.joining(",")))
-                .put(LIST_COLUMN_TYPES, tableColumns.stream().map(HiveColumnHandle::getHiveType).map(HiveType::toString).collect(Collectors.joining(",")))
-                .buildOrThrow();
-        return new PageSourceParameters(factory, length, nowMillis, columnMappings, tableProperties);
+        return new PageSourceParameters(factory, length, nowMillis, columnMappings, getTablePropertiesWithEncoding(tableColumns, BINARY_ENCODING));
```
Comment (linlin-s) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/ion/IonPageSourceSmokeTest.java:349]: Added default Ion encoding to table properties.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/ion/IonPageSourceSmokeTest.java [hunk 10]
```diff
@@ -52,19 +59,28 @@
...
 import static io.trino.type.InternalTypeManager.TESTING_TYPE_MANAGER;
+import static java.util.stream.Collectors.toList;
 
 /**
  * Most basic test to reflect PageSource-fu is wired up correctly.
  */
 public class IonPageSourceSmokeTest
 {
+    // In the Ion binary format, a value stream is always start with binary version marker. This help distinguish Ion binary
+    // data from other formats, including Ion text format.
```
Comment (rmarrowstone) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/ion/IonPageSourceSmokeTest.java:80]:
```suggestion
    // In the Ion binary format, a value stream always starts with binary version marker.
    // This helps distinguish Ion binary data from other formats, including Ion text format.
```

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/ion/IonPageSourceSmokeTest.java [hunk 11]
```diff
@@ -52,19 +59,28 @@
...
+    private static final byte[] BINARY_VERSION_MARKER = {(byte) 0xE0, (byte) 0x01, (byte) 0x00, (byte) 0XEA};
+    private static final String EXPECTED_TEXT = "{foo:3,bar:6}";
```
Comment (rmarrowstone) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/ion/IonPageSourceSmokeTest.java:82]: This is OK as it is, no need to change it, just food for thought: I probably would just assert on "starts with text opening brace", anything else gets brittle fast. This is not so bad in and of itself though.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestHiveFileFormats.java [hunk 12]
```diff
@@ -389,6 +393,54 @@ public void testIon(int rowCount, long fileSizePadding)
...
+                .withColumns(testColumns)
+                .withRowsCount(rowCount)
+                .withFileSizePadding(fileSizePadding)
+                .withTableProperties(ImmutableMap.of(ION_ENCODING_PROPERTY, TEXT_ENCODING))
+                .withFileWriterFactory(fileSystemFactory -> new IonFileWriterFactory(fileSystemFactory, TESTING_TYPE_MANAGER))
+                .isReadableByPageSource(fileSystemFactory -> new IonPageSourceFactory(fileSystemFactory, hiveConfig));
+    }
+
+    @Test(dataProvider = "validRowAndFileSizePadding")
+    public void testInvalidIonEncoding(int rowCount, long fileSizePadding)
```
Comment (rmarrowstone) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestHiveFileFormats.java:420]: This feels pretty "heavyweight" for what you're testing, a lot of setup, a lot of stuff that's extraneous for your test. I suspect the gap you found was in a test over the `IonFileWriterFactory` I would just create that now. We will need to test stuff like null field serialization there anyway.

Comment (rmarrowstone) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestHiveFileFormats.java:420]: Could address this comment later though.

**Review by rmarrowstone — APPROVED** [review]

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestHiveFileFormats.java (hunk 12 above)
Comment (linlin-s) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestHiveFileFormats.java:420]: Yes sounds good, I will add TODO and address this in the later PR.

**Review by desaikd — APPROVED** [review]
